clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -935,18 +935,18 @@ public void handle(HeartbeatResponse heartbeatResponse, RequestFuture<Void> futu
             } else if (error == Errors.REBALANCE_IN_PROGRESS) {
                 log.info("Attempt to heartbeat failed since group is rebalancing");
                 requestRejoin();
-                future.raise(Errors.REBALANCE_IN_PROGRESS);
+                future.raise(error);
             } else if (error == Errors.ILLEGAL_GENERATION) {
                 log.info("Attempt to heartbeat failed since generation {} is not current", generation.generationId);
                 resetGeneration();
-                future.raise(Errors.ILLEGAL_GENERATION);
+                future.raise(error);
             } else if (error == Errors.FENCED_INSTANCE_ID) {
                 log.error("Received fatal exception: group.instance.id gets fenced");
                 future.raise(error);
             } else if (error == Errors.UNKNOWN_MEMBER_ID) {
                 log.info("Attempt to heartbeat failed for since member id {} is not valid.", generation.memberId);
                 resetGeneration();
-                future.raise(Errors.UNKNOWN_MEMBER_ID);
+                future.raise(error);
             } else if (error == Errors.GROUP_AUTHORIZATION_FAILED) {
                 future.raise(new GroupAuthorizationException(groupId));
```

Comment:
Could we raise the error after else block in L954 and only switch it in the last two cases:
```
 } else if (error == Errors.GROUP_AUTHORIZATION_FAILED) {
                error = new GroupAuthorizationException(groupId));
} else {
  error = new KafkaException("Unexpected error in heartbeat response: " + error.message());
}
future.raise(error);
```

Reply:
I feel it does not worth this cleanup for readability since we need another hierarchy of `else` covering the error cases..